src/imitation/util/networks.py

```diff
@@ -117,6 +117,57 @@ def forward(self, x: th.Tensor) -> th.Tensor:
         return (x - self.running_mean) / th.sqrt(self.running_var + self.eps)
 
 
+class EMARunningNorm(RunningNorm):
+    """Similar to RunningNorm but uses an exponentially weighting."""
+
+    running_mean: th.Tensor
+    running_var: th.Tensor
+
+    def __init__(
+        self,
+        num_features: int,
+        lamb: float = 0.05,
+        r: float = 0.05,
+        eps: float = 1e-5,
+    ):
+        """Builds EMARunningNorm.
+
+        Args:
+            num_features: Number of features; the length of the non-batch dim.
+            lamb: value between (0, 1) update rate for the mean
+            r: value between (0, 1) update rate for the variance
+            eps: small constant for for numerical stability.
+        """
+        super().__init__(num_features, eps=eps)
+        self.lamb = lamb
+        self.r = r
+
+    def update_stats(self, batch: th.Tensor) -> None:
+        """Update `self.running_mean` and `self.running_var`.
+
+        Uses Macgregor & Harris (1993), "The Exponentially Weighted Moving Variance".
+
+        Args:
+            batch: A batch of data to use to update the running mean and variance.
+        """
+        if len(batch.shape) == 1:  # Assume list is a batch of 1-d vectors
+            batch = batch.unsqueeze(-1)
+
+        b_size = batch.shape[0]
+        b_mean = th.mean(batch, dim=0)
+        self.running_mean *= 1 - self.lamb
+        self.running_mean += b_mean * self.lamb
+
+        b_var = (
+            th.sum(th.square(self.running_mean.unsqueeze(0) - batch), dim=0) / b_size
+        )
```

Comment:
I think I like the derivations in [Incremental calculation of weighted mean and variance](https://fanf2.user.srcf.net/hermes/doc/antiforgery/stats.pdf) better than “The Exponentially Weighted Moving Variance”. Let me see if I can quickly work out how to take into account the batching.

Reply:
I am convinced that using the old running mean is probably more correct, so I'm going to switch the current implementation over to doing that. The derivations in [this paper](https://fanf2.user.srcf.net/hermes/doc/antiforgery/stats.pdf) are extremely difficult for me to follow in a few places, and thus I'm not 100% sure the way I'm generalizing to batches is correct. I really doubt this will affect the performance of the system at all, so I'm not going to spend any more time trying to derive the incremental update from first principles.

Reply:
This seems relevant https://openreview.net/pdf?id=SkgGjRVKDS

Reply:
I have switched over to an implementation that does take batches into account and simply and updates incrementally. Howerver, it is a bit slower,
```bash
# New implementation
$ python -m timeit -s 'import torch; from imitation.util import networks; norm = networks.EMANorm(256); a = torch.randn(64, 256)' 'norm(a)'
> 200 loops, best of 5: 1.19 msec per loop
# Old implimentation
$ python -m timeit -s 'import torch; from imitation.util import networks; norm = networks.EMANorm(256); a = torch.randn(64, 256)' 'norm(a)'    
5000 loops, best of 5: 66.3 usec per loop
# Running mean implementation in use now
$ python -m timeit -s 'import torch; from imitation.util import networks; norm = networks.RunningNorm(256); a = torch.randn(64, 256)' 'norm(a)'
2000 loops, best of 5: 182 usec per loop
```